### ISSUE TITLE
Add Parametric Morphism on ∀ x y, P x → P y → equ x  y → equ (f x) (f y)

### ISSUE THREAD

**Issue by roglo** [coq/coq#7708]
#### Description of the problem

When we have an equivalence relation "equ" and a theorem saying
      ∀ x y, equ x  y → equ (f x) (f y)
we can use "Add Parametric Morphism : (f) ..." to be able to use rewrite. I often do it.

But here, my theorem is, for some predicate P
      ∀ x y, P x → P y → equ x  y → equ (f x) (f y)
In other words my morphism has a condition. Is is possible to use "Add Parametric Morphism" in that case? How?

I did not find any solution. Is this feature existing in Coq?

I have that in my code
    https://github.com/roglo/coq_homol_algeb/blob/5540a868dfd0da41a6529d9407a7a5161278f09f/AbGroup.v
From lines 107 to 131, there are Morphisms which work.
From lines 140 to 176, an attempt to make it work in my case. The code is commented but it compiles. Its drawback is that is does not allow me to use rewrite.

I asked my fellow Theo Zimmerman and he recommended me to ask here rather than in Coq Club, because he thinks this is missing in Coq. Is he right?
#### Description of the problem

When we have an equivalence relation "equ" and a theorem saying
      ∀ x y, equ x  y → equ (f x) (f y)
we can use "Add Parametric Morphism : (f) ..." to be able to use rewrite. I often do it.

But here, my theorem is, for some predicate P
      ∀ x y, P x → P y → equ x  y → equ (f x) (f y)
In other words my morphism has a condition. Is is possible to use "Add Parametric Morphism" in that case? How?

I did not find any solution. Is this feature existing in Coq?

I have that in my code
    https://github.com/roglo/coq_homol_algeb/blob/5540a868dfd0da41a6529d9407a7a5161278f09f/AbGroup.v
From lines 107 to 131, there are Morphisms which work.
From lines 140 to 176, an attempt to make it work in my case. The code is commented but it compiles. Its drawback is that is does not allow me to use rewrite.

I asked my fellow Theo Zimmerman and he recommended me to ask here rather than in Coq Club, because he thinks this is missing in Coq. Is he right?

**Comment by ppedrot** [comment]
`Add Parametric Morphism` is deprecated, you should declare a `Proper` instance instead.

**Comment by Zimmi48** [comment]
@ppedrot That's not actually true. `Add Parametric Morphism` is just a wrapper around a `Proper` instance, true, but it is not deprecated. The refman chapter https://coq.inria.fr/refman/addendum/generalized-rewriting.html#adding-new-relations-and-morphisms documents this command. And it was never marked as deprecated anywhere. In any case, what is used on the surface doesn't change that the current generalized rewriting feature is missing support for generating side-conditions to be left to the user to prove.

**Comment by ppedrot** [comment]
@Zimmi48 Well, at least it is morally deprecated, even though not documented nor implemented as such. @ejgallego tried to do that formally, but AFAIR he discarded the PR for a lack of support to port the current code. Anyway, it's just a (incomplete) wrapper indeed.

> the current generalized rewriting feature is missing support for generating side-conditions to be left to the user to prove.

Really?

**Comment by ppedrot** [comment]
> the current generalized rewriting feature is missing support for generating side-conditions to be left to the user to prove.

What do you mean by that? The following works:
```
Require Import Setoid.

Lemma R : 0 = 1 -> True <-> False.
Proof.
intros [=].
Qed.

Goal (0 = 0 -> True) <-> (1 = 1 -> False).
rewrite R.

2 subgoals
______________________________________(1/2)
(0 = 0 -> False) <-> (1 = 1 -> False)
______________________________________(2/2)
0 = 1
```

**Comment by JasonGross** [comment]
@ppedrot the lemma you're rewriting with can have side conditions, but the all side conditions of the Proper instances must be resolved by typeclass resolution.  (It might work to explicitly add hints to `shelve` the side conditions and then `unshelve rewrite`; I'm not sure.)

**Comment by Zimmi48** [comment]
I was not talking about this kind of side conditions (side conditions of a lemma that you want to rewrite with) but about side conditions of "morphism" / "proper instances" during the proof search.

The trick suggested by @JasonGross might work indeed, I've used similar ones in similar contexts in the past. But I thought this deserved a proper feature request (no pun intended) and that's why I suggested that Daniel open this issue.

**Comment by JasonGross** [comment]
And if the side conditions are on the elements selected for rewriting, I don't think the `Proper` lemme can even be phrased inn a way that it'll be picked up.  I'd want something like
```coq
Proper ((fun x y => equ x y /\ (P x /\ P y)) ==> equ) f
```
to be picked up when I rewrite with an `equ` lemma.  And if it's to leave over the side conditions for just the chosen variables... I don't think the system has enough information to do that by the time it hits tc resolution, because it's already dropped the occurrences it's chosen in order to phrase the Proper goal to kick off typeclass search.  I don't see an easy way to do this without massively redesigning the system, though it might work to have a special "with side conditions" relation and declare every relation a subrelation of itself-with-trivial-side-conditions, and kick off the tc search with side-conditions...

**Comment by roglo** [comment]
@mattam82 ?
